website/project/model.py

```diff
@@ -3621,81 +3681,97 @@ class DraftRegistration(AddonModelMixin, StoredObject):
+        if meta_schema:
+            schema = meta_schema.schema
+            if not self.registration_schema:
+                flags = schema.get('flags', {})
+                for flag, value in flags.iteritems():
+                    self.flags[flag] = value
+
+    @property
+    def requires_approval(self):
+        return self.registration_schema.requires_approval
+
+    @property
+    def is_pending_review(self):
+        return self.approval.is_pending_approval if (self.requires_approval and self.approval) else False
+
+    @property
+    def is_approved(self):
+        if self.requires_approval and not self.approval:
+            return False
+        elif self.requires_approval and self.approval:
+            return self.approval.is_approved
+        else:
+            return True
 
     def update_metadata(self, metadata, save=True):
-        # TODO: uncommnet to disallow editing drafts that are approved or pending approval
+        # TODO(barbour-em): delete or implement in schema
         # if self.is_pending_review or self.is_approved:
         #    raise NodeStateError('Cannot edit while this draft is being reviewed')
+
         changes = []
-        for key, value in metadata.iteritems():
-            old_value = self.registration_metadata.get(key)
+        for question_id, value in metadata.iteritems():
+
+            old_value = self.registration_metadata.get(question_id)
+
+            if old_value:
+                old_comments = old_value.get('comments', [])
+                new_comments = value.get('comments', [])
+
+                #  we are using the `created` attribute sort of as a primary key
+                old_comment_ids = [comment['created'] for comment in old_comments]
+
+                # Handle comment conflicts
+                for old_comment in old_comments:
+                    for new_comment in new_comments:
+                        new_id = new_comment.get('created', [])
+                        # if the primary key is already in use and the old comment is more recent,
+                        if new_id in old_comment_ids and old_comment['lastModified'] > new_comment['lastModified']:
+                            # use the old one instead
+                            loc = new_comments.index(new_comment)
+                            new_comments[loc] = old_comment
+
             if not old_value or old_value.get('value') != value.get('value'):
-                changes.append(key)
+                changes.append(question_id)
+
         self.registration_metadata.update(metadata)
+
         if save:
             self.save()
-        # TODO: uncomment to nullify approval state if edited
-        # project_signals.draft_edited.send(self, changes)
-        # self.after_edit(changes)
+        project_signals.draft_edited.send(changes)
+        self.after_edit(changes)
         return changes
 
     def before_edit(self, auth):
+        # TODO(samchrisinger): Make sure we still need this
```

Comment:
If these messages can be embedded in the schema itself, this can all be done client side.